File: test/feature/rappels/domain/auto_notification_stress_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_yuztoo/feature/client_notification/domain/promotion_segment_matching.dart';
import 'package:flutter_yuztoo/feature/rappels/domain/auto_notification_triggers.dart';
import 'package:flutter_yuztoo/feature/rappels/presentation/widgets/trigger_grid.dart';

/// Hard tests — Dart/TS contract + segment parity for auto-notifications.
void main() {
  group('HARD: trigger contract Flutter ↔ Cloud Functions', () {
    test('triggerLabels length matches grid count', () {
      expect(triggerLabels.length, 11);
    });

    test('every trigger label is non-empty and unique', () {
      expect(triggerLabels.toSet().length, triggerLabels.length);
      for (final t in triggerLabels) {
        expect(t.trim(), isNotEmpty);
      }
    });

    test('wired triggers set is consistent with visitDetected exception', () {
      for (final t in AutoNotificationTriggers.triggerLabels) {
        if (t == AutoNotificationTriggers.visitDetected) {
          expect(AutoNotificationTriggers.isWiredInCloud(t), isFalse);
        } else {
          expect(AutoNotificationTriggers.isWiredInCloud(t), isTrue);
        }
      }
    });
  });

  group('HARD: segment matrix — promotion targeting (Dart)', () {
    const segments = ['vip', 'habitue', 'nouveau', 'inactif'];
    const targetsList = <List<String>>[
      ['vip'],
      ['soutien'],
      ['habitue'],
      ['nouveau'],
      ['inactif'],
      ['abonne'],
      ['vip', 'nouveau'],
      ['soutien', 'habitue'],
    ];

    for (final seg in segments) {
      for (final targets in targetsList) {
        test('seg=$seg targets=$targets', () {
          final match = promotionSegmentMatchesTarget(seg, targets);
          final viaSoutien = targets.contains('soutien') &&
              (seg == 'vip' || seg == 'habitue');
          // abonne is UI-only; Dart matcher does not map abonne → nouveau
          final viaDirect = targets.contains(seg);
          expect(match, viaSoutien || viaDirect);
        });
      }
    }

    test('empty targets → no match (promo "Tous" uses empty list upstream)', () {
      for (final seg in segments) {
        expect(promotionSegmentMatchesTarget(seg, []), isFalse);
      }
    });

    test('abonne target does not match nouveau in Dart (CF normalizes abonne)', () {
      expect(promotionSegmentMatchesTarget('nouveau', ['abonne']), isFalse);
    });
  });

  group('HARD: birthday / inactive business rules (mirrors CF)', () {
    bool birthdayAllowed({
      required String? dob,
      required String todayMd,
      required int todayYear,
      required int? lastSentYear,
    }) {
      if (dob == null || dob.isEmpty) return false;
      final dobMd = dob.length >= 5 ? dob.substring(dob.length - 5) : dob;
      if (dobMd != todayMd) return false;
      return lastSentYear != todayYear;
    }

    test('no DOB → no birthday notif', () {
      expect(
        birthdayAllowed(dob: null, todayMd: '05-07', todayYear: 2026, lastSentYear: null),
        isFalse,
      );
    });

    test('double send same year blocked', () {
      expect(
        birthdayAllowed(
          dob: '1990-05-07',
          todayMd: '05-07',
          todayYear: 2026,
          lastSentYear: 2026,
        ),
        isFalse,
      );
    });

    bool inactiveAllowed({
      required double daysSinceVisit,
      required int? daysSinceLastNotif,
    }) {
      if (daysSinceVisit < 60) return false;
      if (daysSinceLastNotif == null) return true;
      return daysSinceLastNotif >= 30;
    }

    test('inactive spam guard', () {
      expect(inactiveAllowed(daysSinceVisit: 90, daysSinceLastNotif: 5), isFalse);
      expect(inactiveAllowed(daysSinceVisit: 90, daysSinceLastNotif: 30), isTrue);
    });
  });
}

File: functions/src/__tests__/auto_notification_stress.test.ts
/**
 * Hard / stress tests for auto-notification production logic.
 * Simulates double cron runs, legacy data, segment matrix, and toggle edge cases.
 */
import {
  AUTO_NOTIFICATION_TRIGGERS,
  autoNotificationSegmentMatches,
  filterEnabledNotificationsForTrigger,
  isBirthdayToday,
  isMerchantAutoNotificationsEnabled,
  shouldSendBirthdayThisYear,
  shouldSendConnectionAnniversary,
  shouldSendInactiveReturn,
  triggersMatch,
  WIRED_AUTO_NOTIFICATION_TRIGGERS,
} from "../auto_notification_core";
import { computeSegment } from "../index";

const doc = (trigger: string, enabled = true) => ({
  data: () => ({ trigger, is_enabled: enabled }),
});

describe("HARD: auto-notification production stability", () => {
  describe("H1 — merchant master toggle must gate everything", () => {
    const cases: Array<[unknown, boolean]> = [
      [undefined, false],
      [{}, true],
      [{ notifications_auto_enabled: true }, true],
      [{ notifications_auto_enabled: false }, false],
      [{ notifications_auto_enabled: 0 }, false],
      [{ notifications_auto_enabled: "false" }, true], // only strict false disables
    ];
    test.each(cases)("merchant data %j → enabled=%s", (data, expected) => {
      expect(
        isMerchantAutoNotificationsEnabled(data as Record<string, unknown> | undefined)
      ).toBe(expected);
    });
  });

  describe("H2 — legacy trigger alias matrix (production Firestore drift)", () => {
    test("legacy birthday alias matches", () => {
      expect(
        triggersMatch("Chaque birthday client", AUTO_NOTIFICATION_TRIGGERS.birthday)
      ).toBe(true);
    });
    test("retired promo trigger does not match birthday", () => {
      expect(
        triggersMatch("Chaque promotion créé", AUTO_NOTIFICATION_TRIGGERS.birthday)
      ).toBe(false);
    });
    test("inactive legacy alias matches", () => {
      expect(
        triggersMatch("Rappel inactivité", AUTO_NOTIFICATION_TRIGGERS.inactiveReturn)
      ).toBe(true);
    });
    test("passage legacy alias matches", () => {
      expect(
        triggersMatch(
          "Après validation fidélité",
          AUTO_NOTIFICATION_TRIGGERS.passageValidated
        )
      ).toBe(true);
    });

    test("filter returns only matching enabled docs among noise", () => {
      const docs = [
        doc("Chaque promotion créé"),
        doc("Chaque birthday client"),
        doc(AUTO_NOTIFICATION_TRIGGERS.birthday),
        doc(AUTO_NOTIFICATION_TRIGGERS.inactiveReturn),
        doc(AUTO_NOTIFICATION_TRIGGERS.birthday, false),
      ];
      const birthday = filterEnabledNotificationsForTrigger(
        docs,
        AUTO_NOTIFICATION_TRIGGERS.birthday
      );
      expect(birthday).toHaveLength(2);
    });
  });

  describe("H3 — birthday idempotency (double cron / retry)", () => {
    const today = new Date("2026-05-07T09:00:00+02:00");
    const todayMD = "05-07";
    const dob = "1992-05-07";

    test("first run on birthday → send", () => {
      expect(
        shouldSendBirthdayThisYear(dob, todayMD, today, undefined)
      ).toBe(true);
    });

    test("second run same day same year → no send", () => {
      expect(
        shouldSendBirthdayThisYear(dob, todayMD, today, 2026)
      ).toBe(false);
    });

    test("next calendar year → send again", () => {
      expect(
        shouldSendBirthdayThisYear(dob, todayMD, today, 2025)
      ).toBe(true);
    });

    test("wrong day never sends even without prior send", () => {
      expect(
        shouldSendBirthdayThisYear(dob, "05-08", today, undefined)
      ).toBe(false);
    });
  });

  describe("H4 — inactive anti-spam (was daily spam in prod)", () => {
    const now = Date.parse("2026-06-01T09:00:00Z");

    test("day 59 inactive → no", () => {
      expect(shouldSendInactiveReturn(59, undefined, now)).toBe(false);
    });
    test("day 60 inactive first time → yes", () => {
      expect(shouldSendInactiveReturn(60, undefined, now)).toBe(true);
    });
    test("day 90 inactive, notified 10 days ago → no", () => {
      const tenDaysAgo = now - 10 * 86400000;
      expect(shouldSendInactiveReturn(90, tenDaysAgo, now)).toBe(false);
    });
    test("day 90 inactive, notified 31 days ago → yes", () => {
      const thirtyOneDaysAgo = now - 31 * 86400000;
      expect(shouldSendInactiveReturn(90, thirtyOneDaysAgo, now)).toBe(true);
    });
  });

  describe("H5 — connection anniversary edge years", () => {
    test("same calendar year as follow → never", () => {
      expect(
        shouldSendConnectionAnniversary("03-01", "03-01", 2026, 2026, undefined)
      ).toBe(false);
    });
    test("second year, not yet sent → yes", () => {
      expect(
        shouldSendConnectionAnniversary("03-01", "03-01", 2025, 2026, undefined)
      ).toBe(true);
    });
    test("second year, already sent this year → no", () => {
      expect(
        shouldSendConnectionAnniversary("03-01", "03-01", 2025, 2026, 2026)
      ).toBe(false);
    });
  });

  describe("H6 — segment filter exhaustive (auto notif parity with promos)", () => {
    const segments = ["vip", "habitue", "nouveau", "inactif"] as const;
    const targets = [
      [],
      ["vip"],
      ["habitue"],
      ["nouveau"],
      ["inactif"],
      ["soutien"],
      ["abonne"],
      ["vip", "nouveau"],
      ["soutien", "nouveau"],
    ];

    for (const seg of segments) {
      for (const t of targets) {
        test(`seg=${seg} targets=${JSON.stringify(t)}`, () => {
          const result = autoNotificationSegmentMatches(seg, t);
          if (t.length === 0) {
            expect(result).toBe(true);
            return;
          }
          const viaSoutien =
            t.includes("soutien") &&
            (seg === "vip" || seg === "habitue");
          const viaAbonne = t.includes("abonne") && seg === "nouveau";
          const viaDirect = t.includes(seg);
          expect(result).toBe(viaSoutien || viaAbonne || viaDirect);
        });
      }
    }
  });

  describe("H7 — segment transition triggers (loyalty → CRM)", () => {
    test("passage 2→3 crosses habitue threshold", () => {
      expect(computeSegment(2, 0)).toBe("nouveau");
      expect(computeSegment(3, 0)).toBe("habitue");
    });
    test("61 days idle → inactif overrides vip passages", () => {
      expect(computeSegment(15, 61)).toBe("inactif");
    });
    test("9→10 passages crosses vip", () => {
      expect(computeSegment(9, 10)).toBe("habitue");
      expect(computeSegment(10, 10)).toBe("vip");
    });
  });

  describe("H8 — leap year birthday window", () => {
    test("Feb 29 DOB on Feb 28 2025", () => {
      const d = new Date("2025-02-28T09:00:00+01:00");
      expect(isBirthdayToday("2000-02-29", "02-28", d)).toBe(true);
    });
    test("Feb 29 DOB on Feb 28 2024 leap year → exact match only", () => {
      const d = new Date("2024-02-28T09:00:00+01:00");
      // 02-28 is not 02-29; leap year doesn't use fallback on 28th
      expect(isBirthdayToday("2000-02-29", "02-28", d)).toBe(false);
    });
    test("Feb 29 DOB on Feb 29 2024", () => {
      const d = new Date("2024-02-29T09:00:00+01:00");
      expect(isBirthdayToday("2000-02-29", "02-29", d)).toBe(true);
    });
  });

  describe("H9 — wired triggers contract vs UI", () => {
    test("every canonical trigger except visitDetected is wired", () => {
      for (const t of Object.values(AUTO_NOTIFICATION_TRIGGERS)) {
        if (t === AUTO_NOTIFICATION_TRIGGERS.visitDetected) {
          expect(WIRED_AUTO_NOTIFICATION_TRIGGERS.has(t)).toBe(false);
        } else {
          expect(WIRED_AUTO_NOTIFICATION_TRIGGERS.has(t)).toBe(true);
        }
      }
    });
  });
});
